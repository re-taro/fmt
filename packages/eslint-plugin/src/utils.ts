import type {
	RuleListener,
	RuleWithMeta,
	RuleWithMetaAndName,
} from "@typescript-eslint/utils/eslint-utils";
import type { RuleContext } from "@typescript-eslint/utils/ts-eslint";
import type { Rule } from "eslint";

export interface RuleModule<T extends readonly unknown[]>
	extends Rule.RuleModule {
	defaultOptions: T;
}

/**
 * Creates reusable function to create rules with default options and docs URLs.
 *
 * @param urlCreator Creates a documentation URL for a given rule name.
 * @returns Function to create a rule with the docs URL format.
 */
function RuleCreator() {
	// This function will get much easier to call when this is merged https://github.com/Microsoft/TypeScript/pull/26349
	// TODO - when the above PR lands; add type checking for the context.report `data` property
	return function createNamedRule<
		TOptions extends readonly unknown[],
		TMessageIds extends string,
	>({
		name,
		meta,
		...rule
	}: Readonly<
		RuleWithMetaAndName<TOptions, TMessageIds>
	>): RuleModule<TOptions> {
		return createRule<TOptions, TMessageIds>({
			meta: {
				...meta,
				docs: {
					...meta.docs,
				},
			},
			...rule,
		});
	};
}

/**
 * Creates a well-typed TSESLint custom ESLint rule without a docs URL.
 *
 * @remarks
 * It is generally better to provide a docs URL function to RuleCreator.
 * @returns Well-typed TSESLint custom ESLint rule.
 */
function createRule<
	TOptions extends readonly unknown[],
	TMessageIds extends string,
>({
	create,
	defaultOptions,
	meta,
}: Readonly<RuleWithMeta<TOptions, TMessageIds>>): RuleModule<TOptions> {
	return {
		create: ((
			context: Readonly<RuleContext<TMessageIds, TOptions>>,
		): RuleListener => {
			const optionsWithDefault = context.options.map((options, index) => {
				return {
					...(defaultOptions[index] || {}),
					...(options || {}),
				};
			}) as unknown as TOptions;
			return create(context, optionsWithDefault);
		}) as any,
		defaultOptions,
		meta: meta as any,
	};
}

export const createEslintRule = RuleCreator() as <
	TOptions extends readonly unknown[],
	TMessageIds extends string,
>({
	name,
	meta,
	...rule
}: Readonly<
	RuleWithMetaAndName<TOptions, TMessageIds>
>) => RuleModule<TOptions>;
