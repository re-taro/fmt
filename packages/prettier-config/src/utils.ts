import { createRequire } from "node:module";

import type { Config } from "./types";

export const require = createRequire(import.meta.url);

export const mergeConfig = (...configs: Config[]): Config =>
	configs.reduce<Config>(
		(acc, config) => ({
			...acc,
			...config,
			overrides: [...(acc.overrides ?? []), ...(config.overrides ?? [])],
			plugins: [...(acc.plugins ?? []), ...(config.plugins ?? [])],
		}),
		{},
	);
