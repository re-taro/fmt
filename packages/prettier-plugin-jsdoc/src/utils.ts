import BSearch from "binary-searching";
import type { Options, ParserOptions, Plugin } from "prettier";
import { format } from "prettier";

import { TAGS_DEFAULT } from "./roles";
import type { AllOptions, Token } from "./types";

const convertToModernType = (oldType: string): string =>
  withoutStrings(oldType, (type) => {
    type = type.trim();

    type = type.replace(/\.</g, "<");

    type = type.replace(/\*/g, " any ");

    type = type
      .replace(/^\?\s*(\w+)$/, "$1 | null")
      .replace(/^(\w+)\s*\?$/, "$1 | null");

    let changed = true;
    while (changed) {
      changed = false;
      type = type.replace(
        /(^|[^$\w\xA0-\uFFFF])Array\s*<((?:[^<>=]|=>|=(?!>)|<(?:[^<>=]|=>|=(?!>))+>)+)>/g,
        (_, prefix, inner) => {
          changed = true;

          return `${prefix}(${inner})[]`;
        },
      );
    }

    return type;
  });

/**
 * Given a valid TS type expression, this will replace all string literals in
 * the type with unique identifiers. The modified type expression will be passed
 * to the given map function. The unique identifiers in the output if the map
 * function will then be replaced with the original string literals.
 *
 * This allows the map function to do type transformations without worrying
 * about string literals.
 *
 * @param type
 * @param mapFn
 */
function withoutStrings(type: string, mapFn: (type: string) => string): string {
  const strings: string[] = [];
  let modifiedType = type.replace(
    /(["'])(?:\\(?:\r\n|[\s\S])|(?!\1)[^\\\r\n])*\1/g,
    (m) => {
      strings.push(m);

      return `String$${strings.length - 1}$`;
    },
  );

  if (modifiedType.includes("`")) {
    return type;
  }

  modifiedType = mapFn(modifiedType);

  return modifiedType.replace(/String\$(\d+)\$/g, (_, index) => strings[index]);
}

async function formatType(type: string, options?: Options): Promise<string> {
  try {
    const TYPE_START = "type name = ";

    let pretty = type;

    let rest = false;
    if (pretty.startsWith("...")) {
      rest = true;
      pretty = `(${pretty.slice(3)})[]`;
    }

    pretty = await format(`${TYPE_START}${pretty}`, {
      ...options,
      parser: "typescript",
      plugins: [],
      filepath: "file.ts",
    });
    pretty = pretty.slice(TYPE_START.length);

    pretty = pretty
      .replace(/^\s*/g, "")
      .replace(/[;\n]*$/g, "")
      .replace(/^\|/g, "")
      .trim();

    if (rest) {
      pretty = `...${pretty.replace(/\[\s*\]$/, "")}`;
    }

    return pretty;
  } catch {
    return type;
  }
}

const numberOfAStringInString = (string: string, search: string | RegExp) =>
  (string.match(new RegExp(search, "g")) ?? []).length;

function addStarsToTheBeginningOfTheLines(
  originalComment: string,
  comment: string,
  options: AllOptions,
): string {
  if (
    (options.jsdocCommentLineStrategy === "singleLine" &&
      numberOfAStringInString(comment.trim(), "\n") === 0) ||
    (options.jsdocCommentLineStrategy === "keep" &&
      numberOfAStringInString(originalComment, "\n") === 0)
  ) {
    return `* ${comment.trim()} `;
  }

  return `*${comment.replace(/(?:\n(?!$))/g, "\n * ")}\n `;
}

function capitalizer(str: string): string {
  if (!str) {
    return str;
  }

  if (/^https?:\/\//i.test(str)) {
    return str;
  }

  if (str.startsWith("- ")) {
    return str.slice(0, 2) + capitalizer(str.slice(2));
  }

  return str[0].toUpperCase() + str.slice(1);
}

/**
 * Detects the line ends of the given text.
 *
 * If multiple line ends are used, the most common one will be returned.
 *
 * If the given text is a single line, "lf" will be returned.
 *
 * @param text
 */
function detectEndOfLine(text: string): "cr" | "crlf" | "lf" {
  const counter = {
    "\r": 0,
    "\r\n": 0,
    "\n": 0,
  };

  const lineEndPattern = /\r\n?|\n/g;
  let m;
  // eslint-disable-next-line no-cond-assign
  while ((m = lineEndPattern.exec(text))) {
    counter[m[0] as keyof typeof counter]++;
  }

  const cr = counter["\r"];
  const crlf = counter["\r\n"];
  const lf = counter["\n"];
  const max = Math.max(cr, crlf, lf);

  if (lf === max) {
    return "lf";
  } else if (crlf === max) {
    return "crlf";
  } else {
    return "cr";
  }
}

/**
 * Returns the index of a token within the given token array.
 *
 * This method uses binary search using the token location.
 *
 * @param tokens
 * @param token
 */
const findTokenIndex = (tokens: Token[], token: Token): number =>
  BSearch.eq(tokens, token, (a, b) =>
    a.loc.start.line === b.loc.start.line
      ? a.loc.start.column - b.loc.start.column
      : a.loc.start.line - b.loc.start.line,
  );

async function formatCode(
  result: string,
  beginningSpace: string,
  options: AllOptions,
): Promise<string> {
  const { printWidth, jsdocKeepUnParseAbleExampleIndent } = options;

  if (
    result
      .split("\n")
      .slice(1)
      .every((v) => !v.trim() || v.startsWith(beginningSpace))
  ) {
    result = result.replace(
      new RegExp(`\n${beginningSpace.replace(/[\t]/g, "[\\t]")}`, "g"),
      "\n",
    );
  }

  try {
    let formattedExample = "";
    const examplePrintWith = printWidth - 4;

    formattedExample = await (result.trim().startsWith("{")
      ? format(result || "", {
        ...options,
        parser: "json",
        printWidth: examplePrintWith,
      })
      : format(result || "", {
        ...options,
        printWidth: examplePrintWith,
      }));

    result = formattedExample.replace(/(^|\n)/g, `\n${beginningSpace}`);
  } catch {
    result = `\n${result
      .split("\n")
      .map(
        (l) =>
          `${beginningSpace}${jsdocKeepUnParseAbleExampleIndent ? l : l.trim()
          }`,
      )
      .join("\n")}\n`;

    result = result.replace(/^\n[\s]+\n/g, "\n");
  }

  return result;
}

function findPluginByParser(parserName: string, options: ParserOptions) {
  const tsPlugin = options.plugins.find(
    (plugin) =>
      typeof plugin === "object" &&
      (plugin as any).name &&
      plugin.parsers &&
      // eslint-disable-next-line no-prototype-builtins
      plugin.parsers.hasOwnProperty(parserName),
  ) as Plugin | undefined;

  return !tsPlugin ||
    (tsPlugin as any).name === "@re-taro/prettier-plugin-jsdoc" ||
    // eslint-disable-next-line no-prototype-builtins
    tsPlugin.parsers?.hasOwnProperty("jsdoc-parser")
    ? undefined
    : tsPlugin.parsers?.[parserName];
}

const isDefaultTag = (tag: string): boolean => TAGS_DEFAULT.includes(tag);

export {
  addStarsToTheBeginningOfTheLines,
  capitalizer,
  convertToModernType,
  detectEndOfLine,
  findPluginByParser,
  findTokenIndex,
  formatCode,
  formatType,
  isDefaultTag,
};
