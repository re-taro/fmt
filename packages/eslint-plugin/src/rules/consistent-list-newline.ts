import type { RuleFixer, RuleListener } from "@typescript-eslint/utils/ts-eslint";
import type { TSESTree } from "@typescript-eslint/utils";
import { createEslintRule } from "../utils";

const RULE_NAME = "consistent-list-newline";
type MessageIds = "shouldWrap" | "shouldNotWrap";
type Options = [{
	ArrayExpression?: boolean;
	ArrowFunctionExpression?: boolean;
	CallExpression?: boolean;
	ExportNamedDeclaration?: boolean;
	FunctionDeclaration?: boolean;
	FunctionExpression?: boolean;
	ImportDeclaration?: boolean;
	NewExpression?: boolean;
	ObjectExpression?: boolean;
	TSInterfaceDeclaration?: boolean;
	TSTupleType?: boolean;
	TSTypeLiteral?: boolean;
	TSTypeParameterDeclaration?: boolean;
	TSTypeParameterInstantiation?: boolean;
	ObjectPattern?: boolean;
	ArrayPattern?: boolean;
}];

const rule = createEslintRule<Options, MessageIds>({
	create: (context, [options = {}] = [{}]) => {
		function removeLines(fixer: RuleFixer, start: number, end: number) {
			const range = [start, end] as const;
			const code = context.sourceCode.text.slice(...range);
			return fixer.replaceTextRange(range, code.replace(/(\r\n|\n)/g, ""));
		}

		function check(
			node: TSESTree.Node,
			children: (TSESTree.Node | null)[],
			nextNode?: TSESTree.Node,
		) {
			const items = children.filter(Boolean) as TSESTree.Node[];
			if (items.length === 0)
				return;

			// Look for the opening bracket, we first try to get the first token of the parent node
			// and fallback to the token before the first item
			let startToken = ["CallExpression", "NewExpression"].includes(node.type)
				? undefined
				: context.sourceCode.getFirstToken(node);
			if (startToken?.type !== "Punctuator")
				startToken = context.sourceCode.getTokenBefore(items[0]);

			const endToken = context.sourceCode.getTokenAfter(items[items.length - 1]);
			const startLine = startToken!.loc.start.line;

			if (startToken!.loc.start.line === endToken!.loc.end.line)
				return;

			let mode: "inline" | "newline" | null = null;
			let lastLine = startLine;

			items.forEach((item, idx) => {
				if (mode == null) {
					mode = item.loc.start.line === lastLine ? "inline" : "newline";
					lastLine = item.loc.end.line;
					return;
				}

				const currentStart = item.loc.start.line;

				if (mode === "newline" && currentStart === lastLine) {
					context.report({
						data: {
							name: node.type,
						},
						*fix(fixer) {
							yield fixer.insertTextBefore(item, "\n");
						},
						messageId: "shouldWrap",
						node: item,
					});
				}
				else if (mode === "inline" && currentStart !== lastLine) {
					const lastItem = items[idx - 1];
					if (context.sourceCode.getCommentsBefore(item).length > 0)
						return;
					context.report({
						data: {
							name: node.type,
						},
						*fix(fixer) {
							yield removeLines(fixer, lastItem!.range[1], item.range[0]);
						},
						messageId: "shouldNotWrap",
						node: item,
					});
				}

				lastLine = item.loc.end.line;
			});

			const endRange = nextNode
				? Math.min(
					context.sourceCode.getTokenBefore(nextNode)!.range[0],
					node.range[1],
				)
				: node.range[1];
			const endLoc = context.sourceCode.getLocFromIndex(endRange);

			const lastItem = items[items.length - 1]!;
			if (mode === "newline" && endLoc.line === lastLine) {
				context.report({
					data: {
						name: node.type,
					},
					*fix(fixer) {
						yield fixer.insertTextAfter(lastItem, "\n");
					},
					messageId: "shouldWrap",
					node: lastItem,
				});
			}
			else if (mode === "inline" && endLoc.line !== lastLine) {
				// If there is only one multiline item, we allow the closing bracket to be on the a different line
				if (items.length === 1 && items[0].loc.start.line !== items[1]?.loc.start.line)
					return;
				if (context.sourceCode.getCommentsAfter(lastItem).length > 0)
					return;

				const content = context.sourceCode.text.slice(lastItem.range[1], endRange);
				if (content.includes("\n")) {
					context.report({
						data: {
							name: node.type,
						},
						*fix(fixer) {
							yield removeLines(fixer, lastItem.range[1], endRange);
						},
						messageId: "shouldNotWrap",
						node: lastItem,
					});
				}
			}
		}

		const listenser = {
			ArrayExpression: (node) => {
				check(node, node.elements);
			},
			ArrayPattern(node) {
				check(node, node.elements);
			},
			ArrowFunctionExpression: (node) => {
				if (node.params.length <= 1)
					return;
				check(
					node,
					node.params,
					node.returnType || node.body,
				);
			},
			CallExpression: (node) => {
				check(node, node.arguments);
			},
			ExportNamedDeclaration: (node) => {
				check(node, node.specifiers);
			},
			FunctionDeclaration: (node) => {
				check(
					node,
					node.params,
					node.returnType || node.body,
				);
			},
			FunctionExpression: (node) => {
				check(
					node,
					node.params,
					node.returnType || node.body,
				);
			},
			ImportDeclaration: (node) => {
				check(node, node.specifiers);
			},
			NewExpression: (node) => {
				check(node, node.arguments);
			},
			ObjectExpression: (node) => {
				check(node, node.properties);
			},
			ObjectPattern(node) {
				check(node, node.properties, node.typeAnnotation);
			},
			TSInterfaceDeclaration: (node) => {
				check(node, node.body.body);
			},
			TSTupleType: (node) => {
				check(node, node.elementTypes);
			},
			TSTypeLiteral: (node) => {
				check(node, node.members);
			},
			TSTypeParameterDeclaration(node) {
				check(node, node.params);
			},
			TSTypeParameterInstantiation(node) {
				check(node, node.params);
			},
		} satisfies RuleListener;

		type KeysListener = keyof typeof listenser;
		type KeysOptions = keyof Options[0];

		// Type assertion to check if all keys are exported
		exportType<KeysListener, KeysOptions>();
		exportType<KeysOptions, KeysListener>();

		(Object.keys(options) as KeysOptions[]).forEach((key) => {
			if (options[key] === false)
				delete listenser[key];
		});

		return listenser;
	},
	defaultOptions: [{}],
	meta: {
		docs: {
			description: "Having line breaks styles to object, array and named imports",
			recommended: "stylistic",
		},
		fixable: "whitespace",
		messages: {
			shouldNotWrap: "Should not have line breaks between items, in node {{name}}",
			shouldWrap: "Should have line breaks between items, in node {{name}}",
		},
		schema: [{
			additionalProperties: false,
			properties: {
				ArrayExpression: { type: "boolean" },
				ArrayPattern: { type: "boolean" },
				ArrowFunctionExpression: { type: "boolean" },
				CallExpression: { type: "boolean" },
				ExportNamedDeclaration: { type: "boolean" },
				FunctionDeclaration: { type: "boolean" },
				FunctionExpression: { type: "boolean" },
				ImportDeclaration: { type: "boolean" },
				NewExpression: { type: "boolean" },
				ObjectExpression: { type: "boolean" },
				ObjectPattern: { type: "boolean" },
				TSInterfaceDeclaration: { type: "boolean" },
				TSTupleType: { type: "boolean" },
				TSTypeLiteral: { type: "boolean" },
				TSTypeParameterDeclaration: { type: "boolean" },
				TSTypeParameterInstantiation: { type: "boolean" },
			} satisfies Record<keyof Options[0], { type: "boolean" }>,
			type: "object",
		}],
		type: "layout",
	},
	name: RULE_NAME,
});

// eslint-disable-next-line unused-imports/no-unused-vars
function exportType<A, B extends A>() {}

export default rule;

if (import.meta.vitest) {
	const { afterAll, describe, expect, it } = import.meta.vitest;
	const { RuleTester } = await import("../vendor/rule-tester/src/RuleTester");

	const valids = [
		"const a = { foo: \"bar\", bar: 2 }",
		"const a = {\nfoo: \"bar\",\nbar: 2\n}",
		"const a = [1, 2, 3]",
		"const a = [\n1,\n2,\n3\n]",
		"import { foo, bar } from \"foo\"",
		"import {\nfoo,\nbar\n} from \"foo\"",
		"const a = [`\n\n`, `\n\n`]",
		"log(a, b)",
		"log(\na,\nb\n)",
		"function foo(a, b) {}",
		"function foo(\na,\nb\n) {}",
		"const foo = (a, b) => {\n\n}",
		"const foo = (a, b): {a:b} => {\n\n}",
		"interface Foo { a: 1, b: 2 }",
		"interface Foo {\na: 1\nb: 2\n}",
		"a\n.filter(items => {\n\n})",
		"new Foo(a, b)",
		"new Foo(\na,\nb\n)",
		"function foo<T = {\na: 1,\nb: 2\n}>(a, b) {}",
		"foo(() =>\nbar())",
		"foo(() =>\nbar()\n)",
		`call<{\nfoo: 'bar'\n}>('')`,
		`
	(Object.keys(options) as KeysOptions[])
	.forEach((key) => {
		if (options[key] === false)
			delete listenser[key]
	})
		`,
		`function fn({ foo, bar }: {\nfoo: 'foo'\nbar: 'bar'\n}) {}`,
		{
			code: "foo(\na, b\n)",
			options: [{ CallExpression: false }],
		},
		{
			code: `
	const a = (
		<div>
			{text.map((item, index) => (
				<p>
				</p>
			))}
		</div>
	)
		`,
			parserOptions: {
				ecmaFeatures: {
					jsx: true,
				},
			},
		},
		`
	export const getTodoList = request.post<
		Params,
		ResponseData,
	>('/api/todo-list')
	`,
		{
			code: `
	function TodoList() {
		const { data, isLoading } = useSwrInfinite(
			(page) => ['/api/todo/list', { page }],
			([, params]) => getToDoList(params),
		)
		return <div></div>
	}`,
			parserOptions: {
				ecmaFeatures: {
					jsx: true,
				},
			},
		},
		`
	bar(
		foo => foo
			? ''
			: ''
	)
			`,
		`
	bar(
		(ruleName, foo) => foo
			? ''
			: ''
	)
		`,
		`
	const a = [
		(1),
		(2)
	];
		`,
		"const a = [(1), (2)];",

	];

	// Check snapshot for fixed code
	const invalid = [
		"const a = {\nfoo: \"bar\", bar: 2 }",
		"const a = {foo: \"bar\", \nbar: 2\n}",
		"const a = [\n1, 2, 3]",
		"const a = [1, \n2, 3\n]",
		"import {\nfoo, bar } from \"foo\"",
		"import { foo, \nbar } from \"foo\"",
		"const a = {foo: \"bar\", \r\nbar: 2\r\n}",
		"log(\na, b)",
		"function foo(\na, b) {}",
		"const foo = (\na, b) => {}",
		"const foo = (\na, b): {\na:b} => {}",
		"const foo = (\na, b): {a:b} => {}",
		"interface Foo {\na: 1,b: 2\n}",
		"type Foo = {\na: 1,b: 2\n}",
		"type Foo = [1,2,\n3]",
		"new Foo(1,2,\n3)",
		"new Foo(\n1,2,\n3)",
		"foo(\n()=>bar(),\n()=>\nbaz())",
		"foo(()=>bar(),\n()=>\nbaz())",
		"foo<X,\nY>(1, 2)",
		"foo<\nX,Y>(\n1, 2)",
		"function foo<\nX,Y>() {}",
		"const {a,\nb\n} = c",
		"const [\na,b] = c",
		"foo(([\na,b]) => {})",
		{
			code: `
	const a = (
		<div>
			{text.map((
				item, index) => (
				<p>
				</p>
			))}
		</div>
	)
		`,
			parserOptions: {
				ecmaFeatures: {
					jsx: true,
				},
			},
		},
		`
	export default antfu({
	},
	{
		foo: 'bar'
	}
		// some comment
		// hello
	)`,
	`
	export default antfu({
	},
	// some comment
	{
		foo: 'bar'
	},
	{
	}
		// hello
	)`,
	];

	RuleTester.afterAll = afterAll;
	RuleTester.it = it;
	RuleTester.itOnly = it.only;
	RuleTester.describe = describe;

	const ruleTester = new RuleTester({
		parser: require.resolve("@typescript-eslint/parser"),
	});

	ruleTester.run(RULE_NAME, rule as any, {
		invalid: invalid.map(i => typeof i === "string"
			? {
					code: i,
					errors: null,
					onOutput: (output: string) => {
						expect(output).toMatchSnapshot();
					},
				}
			: {
					...i,
					errors: null,
					onOutput: (output: string) => {
						expect(output).toMatchSnapshot();
					},
				},
		),
		valid: valids,
	});
}
