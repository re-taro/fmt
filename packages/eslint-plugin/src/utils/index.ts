import type { TSESTree } from "@typescript-eslint/utils";
import { ESLintUtils } from "@typescript-eslint/utils";
import type { RuleWithMetaAndName } from "@typescript-eslint/utils/eslint-utils";
import type { RuleModule } from "@typescript-eslint/utils/ts-eslint";

export type Rule<
	TOptions extends readonly unknown[],
	TMessageIds extends string,
> = RuleModule<TMessageIds, TOptions>;

export const createEslintRule: <
	TOptions extends readonly unknown[],
	TMessageIds extends string,
>({
	name,
	meta,
	...rule
}: Readonly<RuleWithMetaAndName<TOptions, TMessageIds>>) => Rule<
	TOptions,
	TMessageIds
> = ESLintUtils.RuleCreator((ruleName) => ruleName);

export function getPreviousNode(
	node?: TSESTree.Node,
): TSESTree.Node | undefined {
	if (!node) {
		return;
	}
	const { parent } = node;
	if (parent && "body" in parent) {
		const { body } = parent;
		if (!Array.isArray(body)) {
			return;
		}
		// @ts-expect-error
		const index = body.indexOf(node);
		if (index > 0) {
			return body[index - 1];
		}
	}
}
