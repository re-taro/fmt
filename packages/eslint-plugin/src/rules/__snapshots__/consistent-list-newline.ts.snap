// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`consistent-list-newline > invalid > 
\\u0009const a = (
\\u0009\\u0009<div>
\\u0009\\u0009\\u0009{text.map((
\\u0009\\u0009\\u0009\\u0009item, index) => (
\\u0009\\u0009\\u0009\\u0009<p>
\\u0009\\u0009\\u0009\\u0009</p>
\\u0009\\u0009\\u0009))}
\\u0009\\u0009</div>
\\u0009)
\\u0009\\u0009 1`] = `
"
	const a = (
		<div>
			{text.map((
				item, 
index
) => (
				<p>
				</p>
			))}
		</div>
	)
		"
`;

exports[`consistent-list-newline > invalid > 
\\u0009export default re_taro({
\\u0009},
\\u0009// some comment
\\u0009{
\\u0009\\u0009foo: 'bar'
\\u0009},
\\u0009{
\\u0009}
\\u0009\\u0009// hello
\\u0009) 1`] = `
"
	export default re_taro({
	},
	// some comment
	{
		foo: 'bar'
	},	{
	}
		// hello
	)"
`;

exports[`consistent-list-newline > invalid > 
\\u0009export default re_taro({
\\u0009},
\\u0009{
\\u0009\\u0009foo: 'bar'
\\u0009}
\\u0009\\u0009// some comment
\\u0009\\u0009// hello
\\u0009) 1`] = `
"
	export default re_taro({
	},	{
		foo: 'bar'
	}
		// some comment
		// hello
	)"
`;

exports[`consistent-list-newline > invalid > const [
a,b] = c 1`] = `
"const [
a,
b
] = c"
`;

exports[`consistent-list-newline > invalid > const {a,
b
} = c 1`] = `"const {a,b} = c"`;

exports[`consistent-list-newline > invalid > const a = [
1, 2, 3] 1`] = `
"const a = [
1, 
2, 
3
]"
`;

exports[`consistent-list-newline > invalid > const a = [1, 
2, 3
] 1`] = `"const a = [1, 2, 3]"`;

exports[`consistent-list-newline > invalid > const a = {
foo: "bar", bar: 2 } 1`] = `
"const a = {
foo: "bar", 
bar: 2
 }"
`;

exports[`consistent-list-newline > invalid > const a = {foo: "bar", 
bar: 2
} 1`] = `"const a = {foo: "bar", bar: 2}"`;

exports[`consistent-list-newline > invalid > const a = {foo: "bar", \\u000d
bar: 2\\u000d
} 1`] = `"const a = {foo: "bar", bar: 2}"`;

exports[`consistent-list-newline > invalid > const foo = (
a, b) => {} 1`] = `
"const foo = (
a, 
b
) => {}"
`;

exports[`consistent-list-newline > invalid > const foo = (
a, b): {
a:b} => {} 1`] = `
"const foo = (
a, 
b
): {
a:b
} => {}"
`;

exports[`consistent-list-newline > invalid > const foo = (
a, b): {a:b} => {} 1`] = `
"const foo = (
a, 
b
): {a:b} => {}"
`;

exports[`consistent-list-newline > invalid > foo(
()=>bar(),
()=>
baz()) 1`] = `
"foo(
()=>bar(),
()=>
baz()
)"
`;

exports[`consistent-list-newline > invalid > foo(()=>bar(),
()=>
baz()) 1`] = `
"foo(()=>bar(),()=>
baz())"
`;

exports[`consistent-list-newline > invalid > foo(([
a,b]) => {}) 1`] = `
"foo(([
a,
b
]) => {})"
`;

exports[`consistent-list-newline > invalid > foo<
X,Y>(
1, 2) 1`] = `
"foo<
X,
Y
>(
1, 
2
)"
`;

exports[`consistent-list-newline > invalid > foo<X,
Y>(1, 2) 1`] = `"foo<X,Y>(1, 2)"`;

exports[`consistent-list-newline > invalid > function foo(
a, b) {} 1`] = `
"function foo(
a, 
b
) {}"
`;

exports[`consistent-list-newline > invalid > function foo<
X,Y>() {} 1`] = `
"function foo<
X,
Y
>() {}"
`;

exports[`consistent-list-newline > invalid > import {
foo, bar } from "foo" 1`] = `
"import {
foo, 
bar
 } from "foo""
`;

exports[`consistent-list-newline > invalid > import { foo, 
bar } from "foo" 1`] = `"import { foo, bar } from "foo""`;

exports[`consistent-list-newline > invalid > interface Foo {
a: 1,b: 2
} 1`] = `
"interface Foo {
a: 1,
b: 2
}"
`;

exports[`consistent-list-newline > invalid > log(
a, b) 1`] = `
"log(
a, 
b
)"
`;

exports[`consistent-list-newline > invalid > new Foo(
1,2,
3) 1`] = `
"new Foo(
1,
2,
3
)"
`;

exports[`consistent-list-newline > invalid > new Foo(1,2,
3) 1`] = `"new Foo(1,2,3)"`;

exports[`consistent-list-newline > invalid > type Foo = [1,2,
3] 1`] = `"type Foo = [1,2,3]"`;

exports[`consistent-list-newline > invalid > type Foo = {
a: 1,b: 2
} 1`] = `
"type Foo = {
a: 1,
b: 2
}"
`;
